<!DOCTYPE html>
<html
 dir="ltr"
 lang="en"
 >
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content=#ffffff>
    <link rel="stylesheet" href="/assets/css/app.css">
    <link rel="shortcut icon" type="image/png"
           href="/Main_Plots/logo.png" 
    />
    <script defer src="https://unpkg.com/alpinejs@3.9.0/dist/cdn.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css">
    
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bulma-social@1/bin/bulma-social.min.css">
    
    <!-- Begin Jekyll SEO tag v2.8.0 -->
<title>SPARED Dataset | danielaruizl1.github.io</title>
<meta name="generator" content="Jekyll v3.9.5" />
<meta property="og:title" content="SPARED" />
<meta property="og:locale" content="en US" />
<link rel="canonical" href="http://localhost:4000/" />
<meta property="og:url" content="http://localhost:4000/" />
<meta property="og:site name" content="danielaruizl1.github.io" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="SPARED" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebSite","headline":"SPARED","name":"danielaruizl1.github.io","url":"http://localhost:4000/"}</script>
<!-- End Jekyll SEO tag -->

<style>


    section {
        margin-top: 20px;
    }
    h2 {
        font-size: 20px;
        color: #555;
    }
    ul {
        list-style-type: none; /* Remove bullets */
        padding: 0;
    }
    .li_grande {
        margin-bottom: 10px;
        padding: 8px 0;
    }



</style>

</head>


<body>
    <nav class="navbar is-primary " x-data="{ openNav: false }">
    <div class="container">
    <div class="navbar-brand">
        <a href="/" class="navbar-item">
            SpaRED
        </a>
        <a role="button" class="navbar-burger burger" aria-label="menu" aria-expanded="false" data-target="navMenu" :class="{ 'is-active': openNav }" x-on:click="openNav = !openNav">
            <span aria-hidden="true"></span>
            <span aria-hidden="true"></span>
            <span aria-hidden="true"></span>
        </a>
    </div>
    <div class="navbar-menu" id="navMenu" :class="{ 'is-active': openNav }">
        <div class="navbar-start">
            <a href="/metadata.html" class="navbar-item"> Database Metadata</a>
            <a href="/People_Page/people.html" class="navbar-item">People</a> <!-- Enlace a pagina de people -->
            <div class="navbar-item has-dropdown is-hoverable">
                <a class="navbar-link">Details</a>
                <div class="navbar-dropdown">
                    <a href="/Processing_details/dataset_filtering.html" class="navbar-item">Dataset Filtering</a>
                    <a href="/Processing_details/dataset_processing.html" class="navbar-item">Dataset Processing</a>
                    
                </div>
            </div>
        
            
        </div>
    </div>
    <div class="navbar-end">
        
    </div>


    </div>
    </nav>


    <section class="section">                           
        <div class="content">
        <h1> <b>Dataset Proccesing</b></h1>
                <ul>
                    <li class="li_grande"><ul> Apply tpm normalization: 
                        <li>Normalize gene expression levels using Transcripts Per Million (TPM). The GTF files used for this process can be downloaded from these links: 
                            <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.basic.annotation.gtf.gz">gtf_human_file</a> -  
                            <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M33/gencode.vM33.basic.annotation.gtf.gz">gtf_mouse_file</a></li> 
                        <li>Remove genes that are not found in the GTF annotation file.</li>
                        </ul>
                    </li>
                    <li class="li_grande">Transform the data by applying a logarithm base 2 using <a href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.log1p.html"> scanpy.pp.log1p()</a> function.</li>
                    <li class="li_grande">The adaptive median filter is applied to all the slides in the collection. 
                        The maximum window size is defined by a neighborhood of 3 hops, which represents the number of concentric rings in a graph considered when computing the median. 
                        This adaptive median filter is the one proposed in the <a href="https://openaccess.thecvf.com/content/ICCV2023W/CVAMD/html/Mejia_SEPAL_Spatial_Gene_Expression_Prediction_from_Local_Graphs_ICCVW_2023_paper.html">SEPAL paper</a>.
                                        </li>
                    <li class="li_grande">Compute Moran's I for each gene in each slide using <a href="https://squidpy.readthedocs.io/en/stable/api/squidpy.gr.spatial_autocorr.html"> squidpy.gr.spatial_autocorr()</a> function. Then, average Moran's I for each gene across all slides.</li>
                    <li class="li_grande">Filter by Moran's I, keeping the top 128 or 32 genes with the highest Moran's I. This parameter varies depending on each dataset. 
                        For a detailed list of the number of genes in each dataset, please refer to <a href="/metadata.html"> Database Metadata</a>.</li>
                    <li class="li_grande">Perform ComBat batch correction using <a href="https://epigenelabs.github.io/pyComBat/_modules/pycombat.html">pycombat()</a> function.</li>
                    <li class="li_grande">Compute the deviations from the mean expression of each gene.</li>
                </ul>
        </div>
    </section>
    

    <section class="footer">
    <div class="container">
        <div class="content is-small has-text-centered">
            <p class="">Theme built by <a href="https://www.csrhymes.com">C.S. Rhymes</a></p>
        </div>
    </div>
    </section>

    
<script src="/assets/js/app.js" type="text/javascript"></script><!-- footer scripts -->
</body>
</html>
